pyo3-macros-backend/src/pyclass.rs

```diff
@@ -640,6 +770,156 @@ fn impl_enum(
+        variant_cls_pytypeinfos.push(variant_cls_pytypeinfo);
+
+        let variant_new = complex_enum_variant_new(cls, variant)?;
+
+        let (variant_cls_impl, field_getters) = impl_complex_enum_variant_cls(cls, &variant)?;
+        variant_cls_impls.push(variant_cls_impl);
+
+        let pyclass_impl = PyClassImplsBuilder::new(
+            &variant_cls,
+            &variant_args,
+            PyClassMethodsType::Specialization,
+            field_getters,
+            vec![variant_new],
+        )
+        .impl_all()?;
+
+        variant_cls_pyclass_impls.push(pyclass_impl);
+    }
+
+    Ok(quote! {
+        const _: () = {
+            use #krate as _pyo3;
+
+            #pytypeinfo
+
+            #pyclass_impls
+
+            #[doc(hidden)]
+            #[allow(non_snake_case)]
+            impl #cls {}
+
+            #(#variant_cls_zsts)*
+
+            #(#variant_cls_pytypeinfos)*
+
+            #(#variant_cls_pyclass_impls)*
+
+            #(#variant_cls_impls)*
+        };
+    })
+}
+
+fn impl_complex_enum_variant_cls(
+    enum_name: &syn::Ident,
+    variant: &PyClassComplexEnumVariant<'_>,
+) -> Result<(TokenStream, Vec<MethodAndMethodDef>)> {
+    let variant_ident = &variant.ident;
+    let variant_cls = gen_complex_enum_variant_class_ident(enum_name, &variant.ident);
+    let variant_cls_type = parse_quote!(#variant_cls);
+
+    let mut field_names: Vec<Ident> = vec![];
+    let mut fields_with_types: Vec<TokenStream> = vec![];
+    let mut field_getters = vec![];
+    let mut field_getter_impls: Vec<TokenStream> = vec![];
+    for field in &variant.fields {
+        let field_name = field.ident;
+        let field_type = field.ty;
+        let field_with_type = quote! { #field_name: #field_type };
+
+        let field_getter = complex_enum_variant_field_getter(
+            &variant_cls_type,
+            &field_name,
+            field_type,
+            field.span,
+        )?;
+
+        let field_getter_impl = quote! {
+            fn #field_name(slf: _pyo3::PyRef<Self>) -> _pyo3::PyResult<#field_type> {
+                match &*slf.into_super() {
+                    #enum_name::#variant_ident { #field_name, .. } => Ok(#field_name.clone()),
+                    _ => unreachable!("Wrong complex enum variant found in variant wrapper PyClass"),
+                }
+            }
+        };
+
+        field_names.push(field_name.clone());
+        fields_with_types.push(field_with_type);
+        field_getters.push(field_getter);
+        field_getter_impls.push(field_getter_impl);
+    }
+
+    let cls_impl = quote! {
+        #[doc(hidden)]
+        #[allow(non_snake_case)]
+        impl #variant_cls {
+            fn __pymethod_constructor__(py: _pyo3::Python<'_>, #(#fields_with_types)*) -> _pyo3::PyObject {
+                use _pyo3::IntoPy;
+                #enum_name::#variant_ident { #(#field_names)* }.into_py(py)
+            }
```

Comment:
Hmm, thinking about this further, probably you should make this return `PyClassInitializer<#cls>`, based on something like the implementation of the `IntoPy` suggestion I wrote at the bottom of https://github.com/PyO3/pyo3/pull/3582#pullrequestreview-1738372364

Reply:
This also did the trick, thank you! :)